{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c1bb888c-cd9d-4e9b-87de-1432760e5a9e",
   "metadata": {},
   "source": [
    "### RULES\n",
    "Requires format of skipped line between different phylums (ex. empty row above Diatom, Dinoflagellate, etc.)\n",
    "\n",
    "Assumed all {Ochromonas, } are mixotrophs.\n",
    "\n",
    "1. assume everything after \"Unknown flagellates\" is irrelevant (to be deleted)\n",
    "2. diatoms are NOT mixotrophs\n",
    "3. remove all \"[name]-like\" (without genus specified)\n",
    "4. remove all \"[genus name] spp.\" AND \"[genus name] sp.\"\n",
    "5. check \"cysts of\"\n",
    "\n",
    "Status Key--  \n",
    "Confirmed := explicitly in the Mixotroph Database  \n",
    "Unsure (sp. in mdb) := genus in Mixotroph Database lists \"[genus name] sp.\" (ex. Ochromonas sp. for Ochromonas danica)  \n",
    "Unsure (inexact name):= LIS name is in a longer Mixotroph Database name or vice versa (ex. Chattonella marina in Chattonella marina var. ovata)   "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2718b71-fbb0-4c73-8972-e0dfaa85136c",
   "metadata": {},
   "source": [
    "### QUESTIONS TO ASK\n",
    "\n",
    "1. Should I be considering \"cysts of Linggulodinium polyedrum\" mixotrophs?\n",
    "2. How should I handle these \"unsure\" cases? - see status key above"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "f0081d70-60c2-4062-a222-7fcfb5fbe6d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from datetime import datetime\n",
    "\n",
    "pd.set_option(\"future.no_silent_downcasting\", True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "5a6f3687-f38c-4db2-95d2-cdd07727f456",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>1</th>\n",
       "      <th>Species Name</th>\n",
       "      <th>Taxonomic Group</th>\n",
       "      <th>AphiaID</th>\n",
       "      <th>Additional notes</th>\n",
       "      <th>Gene markers</th>\n",
       "      <th>PR2 Accession Number</th>\n",
       "      <th>GenBank Accession Number</th>\n",
       "      <th>Reference to sequence</th>\n",
       "      <th>MFT</th>\n",
       "      <th>Evidence of mixoplankton activity</th>\n",
       "      <th>...</th>\n",
       "      <th>REDS</th>\n",
       "      <th>SANT</th>\n",
       "      <th>SARC</th>\n",
       "      <th>SATL</th>\n",
       "      <th>SPSG</th>\n",
       "      <th>SSTC</th>\n",
       "      <th>SUND</th>\n",
       "      <th>TASM</th>\n",
       "      <th>WARM</th>\n",
       "      <th>WTRA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acanthochiasma sp.</td>\n",
       "      <td>Radiolaria</td>\n",
       "      <td>368427</td>\n",
       "      <td>Acantharia</td>\n",
       "      <td>18S_rRNA_nucleus;18S_rRNA_nucleus;18S_rRNA_nuc...</td>\n",
       "      <td>HM103395.1.1099_U;HM103418.1.1104_U;JN811207.1...</td>\n",
       "      <td>HM103395;HM103418;JN811207;GU825020;HM103399;H...</td>\n",
       "      <td>Quaiser,A.. Comparative metagenomics of bathyp...</td>\n",
       "      <td>eSNCM</td>\n",
       "      <td>endosymbionts</td>\n",
       "      <td>...</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acanthometra fusca</td>\n",
       "      <td>Radiolaria</td>\n",
       "      <td>not registered</td>\n",
       "      <td>Acantharia</td>\n",
       "      <td>18S_rRNA_nucleus;18S_rRNA_nucleus;18S_rRNA_nuc...</td>\n",
       "      <td>KC172856.1.1696_U;EU446351.1.1552_U;JN811165.1...</td>\n",
       "      <td>KC172856;EU446351;JN811165</td>\n",
       "      <td>Decelle,J.. Diversity, ecology and biogeochemi...</td>\n",
       "      <td>eSNCM</td>\n",
       "      <td>endosymbionts</td>\n",
       "      <td>...</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Acanthodesmia vinculata</td>\n",
       "      <td>Radiolaria</td>\n",
       "      <td>493675</td>\n",
       "      <td>Acantharia</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>eSNCM</td>\n",
       "      <td>endosymbionts</td>\n",
       "      <td>...</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>15</td>\n",
       "      <td>not recorded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acanthometra pellucida</td>\n",
       "      <td>Radiolaria</td>\n",
       "      <td>235750</td>\n",
       "      <td>Acantharia</td>\n",
       "      <td>18S_rRNA_nucleus;18S_rRNA_nucleus;18S_rRNA_nuc...</td>\n",
       "      <td>JN811196.1.1668_U;JQ697712.1.1693_U;JQ697708.1...</td>\n",
       "      <td>JN811196;JQ697712;JQ697708;JN811190;JQ697711;J...</td>\n",
       "      <td>Decelle,J.. Molecular Phylogeny and Morphologi...</td>\n",
       "      <td>eSNCM</td>\n",
       "      <td>endosymbionts</td>\n",
       "      <td>...</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acanthometron sp.</td>\n",
       "      <td>Radiolaria</td>\n",
       "      <td>391880</td>\n",
       "      <td>Acantharia</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>eSNCM</td>\n",
       "      <td>endosymbionts</td>\n",
       "      <td>...</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 107 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "1             Species Name Taxonomic Group         AphiaID Additional notes  \\\n",
       "0       Acanthochiasma sp.      Radiolaria          368427       Acantharia   \n",
       "1       Acanthometra fusca      Radiolaria  not registered       Acantharia   \n",
       "2  Acanthodesmia vinculata      Radiolaria          493675       Acantharia   \n",
       "3   Acanthometra pellucida      Radiolaria          235750       Acantharia   \n",
       "4        Acanthometron sp.      Radiolaria          391880       Acantharia   \n",
       "\n",
       "1                                       Gene markers  \\\n",
       "0  18S_rRNA_nucleus;18S_rRNA_nucleus;18S_rRNA_nuc...   \n",
       "1  18S_rRNA_nucleus;18S_rRNA_nucleus;18S_rRNA_nuc...   \n",
       "2                                       not recorded   \n",
       "3  18S_rRNA_nucleus;18S_rRNA_nucleus;18S_rRNA_nuc...   \n",
       "4                                       not recorded   \n",
       "\n",
       "1                               PR2 Accession Number  \\\n",
       "0  HM103395.1.1099_U;HM103418.1.1104_U;JN811207.1...   \n",
       "1  KC172856.1.1696_U;EU446351.1.1552_U;JN811165.1...   \n",
       "2                                       not recorded   \n",
       "3  JN811196.1.1668_U;JQ697712.1.1693_U;JQ697708.1...   \n",
       "4                                       not recorded   \n",
       "\n",
       "1                           GenBank Accession Number  \\\n",
       "0  HM103395;HM103418;JN811207;GU825020;HM103399;H...   \n",
       "1                         KC172856;EU446351;JN811165   \n",
       "2                                       not recorded   \n",
       "3  JN811196;JQ697712;JQ697708;JN811190;JQ697711;J...   \n",
       "4                                       not recorded   \n",
       "\n",
       "1                              Reference to sequence    MFT  \\\n",
       "0  Quaiser,A.. Comparative metagenomics of bathyp...  eSNCM   \n",
       "1  Decelle,J.. Diversity, ecology and biogeochemi...  eSNCM   \n",
       "2                                       not recorded  eSNCM   \n",
       "3  Decelle,J.. Molecular Phylogeny and Morphologi...  eSNCM   \n",
       "4                                       not recorded  eSNCM   \n",
       "\n",
       "1 Evidence of mixoplankton activity  ...          REDS          SANT  \\\n",
       "0                     endosymbionts  ...  not recorded  not recorded   \n",
       "1                     endosymbionts  ...  not recorded  not recorded   \n",
       "2                     endosymbionts  ...  not recorded  not recorded   \n",
       "3                     endosymbionts  ...  not recorded  not recorded   \n",
       "4                     endosymbionts  ...  not recorded  not recorded   \n",
       "\n",
       "1          SARC          SATL          SPSG          SSTC          SUND  \\\n",
       "0  not recorded  not recorded  not recorded  not recorded  not recorded   \n",
       "1  not recorded  not recorded  not recorded  not recorded  not recorded   \n",
       "2  not recorded             2             8  not recorded  not recorded   \n",
       "3  not recorded  not recorded  not recorded  not recorded  not recorded   \n",
       "4  not recorded  not recorded  not recorded  not recorded  not recorded   \n",
       "\n",
       "1          TASM          WARM          WTRA  \n",
       "0  not recorded  not recorded  not recorded  \n",
       "1  not recorded  not recorded  not recorded  \n",
       "2  not recorded            15  not recorded  \n",
       "3  not recorded  not recorded  not recorded  \n",
       "4  not recorded  not recorded  not recorded  \n",
       "\n",
       "[5 rows x 107 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdb = pd.read_csv(\"MDB - 3Dec2022.csv\")\n",
    "mdb.columns = mdb.iloc[1]\n",
    "mdb = mdb.drop([0, 1]).reset_index(drop=True)\n",
    "\n",
    "# edit mdb so that species ending in \"sp\" now end in \"sp.\"\n",
    "mdb['Species Name'] = mdb['Species Name'].str.replace(r'sp$', 'sp.', regex=True)\n",
    "\n",
    "mdb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "fabf7b28-8578-44f2-b83f-6ba228a33654",
   "metadata": {},
   "outputs": [],
   "source": [
    "csv_name = \"LIS_2019-Phytoplankton_Final Report Data.xlsx - 2019 LIS phytoplankton count\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "e852c70f-e894-4588-bc06-4a3079f32083",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import and clean LIS data\n",
    "lis = pd.read_csv(f\"inputs/{csv_name}.csv\")\n",
    "original_headers = lis.columns  # save original column headers\n",
    "\n",
    "phylum_ind = lis[lis.iloc[:, 0] == \"Phylum\"].index[0]\n",
    "lis.columns = lis.iloc[phylum_ind]  # reset column headers\n",
    "lis = lis.iloc[phylum_ind+2:].reset_index(drop=True)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "ae223ae9-e76a-4383-ad84-b79487293ab6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove rows after unknown flagellates\n",
    "unknown_flagellates_ind = lis[lis[\"Phylum\"] == \"Unknown flagellates\"].index[0] \n",
    "lis = lis.iloc[:unknown_flagellates_ind]\n",
    "lis = lis.iloc[:lis.last_valid_index()+1]  # remove trailing nan rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "948f15e0-5c49-4fd1-af99-bb7af42bdce8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove rows that contain \"TOTAL\"\n",
    "lis = lis[~lis[\"Phylum\"].str.contains(\"TOTAL\", na=False)].reset_index(drop=True)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "28a87a73-ccc3-4a22-bc8e-f22c68ebcbb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# construct correct phylum column\n",
    "actual_phylum_ind = lis[lis[\"Species\"].isna() & lis[\"Phylum\"].isna()].index + 1\n",
    "lis = lis.rename(columns={\"Phylum\": \"Genus\"}) # rename phylum column to genus\n",
    "lis.insert(0, 'Phylum', lis[\"Genus\"].iloc[actual_phylum_ind])  # reconstruct phylum column\n",
    "lis['Phylum'] = lis['Phylum'].ffill()  # forwardfill phylum\n",
    "\n",
    "lis['Genus'] = lis['Species'].str.split().str[0]  # fill genus using first word of species name\n",
    "\n",
    "lis = lis.dropna(subset=['Species']).reset_index(drop=True) # delete rows with na in Species column\n",
    "\n",
    "SPECIES_COL = 3\n",
    "lis.iloc[:, SPECIES_COL+1:] = lis.iloc[:, SPECIES_COL+1:].replace(\",\", \"\", regex=True).astype(float)  # ensure numerical values are floats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "ca4ca7ba-58a3-4493-9055-961238f4182e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add Status column\n",
    "lis.insert(0, 'Status', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "20575f78-e4dc-498b-8ddb-d07d9140aeb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# store blocks of known mixotroph genuses \n",
    "ochromonas_ind = lis[lis[\"Species\"].str.contains(\"Ochromonas\")].index\n",
    "ochromonas_block = lis.iloc[ochromonas_ind] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "f4a71b12-d2b8-4700-a53c-2c17e8aba73b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove based on hard coded rules (NOT RESETTING INDEX IN ORDER TO ADD BLOCKS BACK CORRECTLY)\n",
    "lis = lis[lis[\"Phylum\"] != \"Diatom\"] # remove all diatoms\n",
    "lis = lis[~lis[\"Species\"].str.contains(\"-like\")] # remove species ending with \"-like\"\n",
    "lis = lis[~lis[\"Species\"].str.contains(\"sp.|spp.\")]  # remove all sp. / spp."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "24ce9749-f358-4b2b-956f-f9e94032cf49",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check \"cysts of\"\n",
    "CYSTS_LEN = len(\"cysts of \")\n",
    "cysts_of = lis[lis[\"Species\"].str.contains(\"cysts of\", regex=False)][\"Species\"].str.slice(CYSTS_LEN)\n",
    "filtered = cysts_of.isin(mdb['Species Name'])\n",
    "lis.loc[filtered[filtered].index, \"Status\"] = \"Confirmed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "c635deda-718b-46a5-bd0f-654d7e504f6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add back stored blocks of known mixotrophs and mark as Confirmed\n",
    "lis = pd.concat([lis, ochromonas_block]).sort_index().drop_duplicates()\n",
    "lis.loc[ochromonas_ind, \"Status\"] = \"Confirmed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "a6f4a0bc-3b23-49ce-b3c9-200cd295a498",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check if (in none status) direct match and mark all Trues as \"Confirmed\"\n",
    "filtered = lis[lis['Status'].isnull()][\"Species\"].isin(mdb['Species Name'])\n",
    "lis.loc[filtered[filtered].index, \"Status\"] = \"Confirmed\"\n",
    "\n",
    "# check (in remaining none status) if the genus has sp. and mark all Trues as \"Unsure (sp. in mdb)\"\n",
    "genus_to_check = lis[lis['Status'].isnull()]['Species'].str.split().str[0].drop_duplicates() + \" sp.\"\n",
    "filtered = genus_to_check.isin(mdb['Species Name'])\n",
    "lis.loc[filtered[filtered].index, \"Status\"] = \"Unsure (sp. mdb)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "4d751f4e-d0e8-4ac5-bd3d-a5344d8ce133",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check (in remaining none status) if the name is contained in the mdb and vice versa and mark all Trues as \"Unsure (inexact name)\"\n",
    "filtered = lis[lis['Status'].isnull()][\"Species\"].apply(lambda x: mdb[\"Species Name\"].str.contains(x, regex=False).any())\n",
    "lis.loc[filtered[filtered].index, \"Status\"] = \"Unsure (inexact name)\"\n",
    "\n",
    "pattern = '|'.join(mdb['Species Name'])\n",
    "filtered = lis[lis['Status'].isnull()][\"Species\"].str.contains(pattern, regex=True)\n",
    "lis.loc[filtered[filtered].index, \"Status\"] = \"Unsure (inexact name)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "96679bc3-4754-4fdb-bf9f-90b4d7b1d240",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop all rows with Status = \"None\"\n",
    "lis = lis.dropna(subset=['Status']).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "5e49344c-1b4f-4762-a127-abe398a00a48",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>1</th>\n",
       "      <th>Status</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>1/3/19</th>\n",
       "      <th>1/3/19</th>\n",
       "      <th>1/3/19</th>\n",
       "      <th>1/7/19</th>\n",
       "      <th>1/7/19</th>\n",
       "      <th>1/7/19</th>\n",
       "      <th>...</th>\n",
       "      <th>12/6/19</th>\n",
       "      <th>12/6/19</th>\n",
       "      <th>12/6/19</th>\n",
       "      <th>12/5/19</th>\n",
       "      <th>12/16/19</th>\n",
       "      <th>12/16/19</th>\n",
       "      <th>12/16/19</th>\n",
       "      <th>12/4/19</th>\n",
       "      <th>12/4/19</th>\n",
       "      <th>12/4/19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td>TOTAL DINOFLAGELLATES</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>17952.0</td>\n",
       "      <td>17600.0</td>\n",
       "      <td>30800.0</td>\n",
       "      <td>8800.0</td>\n",
       "      <td>0</td>\n",
       "      <td>17600.0</td>\n",
       "      <td>...</td>\n",
       "      <td>13288.0</td>\n",
       "      <td>13200.0</td>\n",
       "      <td>2904.0</td>\n",
       "      <td>352.0</td>\n",
       "      <td>0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>0</td>\n",
       "      <td>2904.0</td>\n",
       "      <td>704.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td></td>\n",
       "      <td>TOTAL RAPHIDOPHYTES</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td></td>\n",
       "      <td>TOTAL OCHROPHYTE (DIATOMS EXCLUDED)S</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8800.0</td>\n",
       "      <td>13200.0</td>\n",
       "      <td>8800.0</td>\n",
       "      <td>...</td>\n",
       "      <td>17600.0</td>\n",
       "      <td>17600.0</td>\n",
       "      <td>17600.0</td>\n",
       "      <td>8800.0</td>\n",
       "      <td>22000.0</td>\n",
       "      <td>22000.0</td>\n",
       "      <td>110000.0</td>\n",
       "      <td>22000.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td></td>\n",
       "      <td>TOTAL HAPTOPHYTES</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 244 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "1 Status                                Phylum Genus Species   1/3/19  \\\n",
       "0                        TOTAL DINOFLAGELLATES                17952.0   \n",
       "1                          TOTAL RAPHIDOPHYTES                      0   \n",
       "2         TOTAL OCHROPHYTE (DIATOMS EXCLUDED)S                      0   \n",
       "3                            TOTAL HAPTOPHYTES                      0   \n",
       "\n",
       "1   1/3/19   1/3/19  1/7/19   1/7/19   1/7/19  ...  12/6/19  12/6/19  12/6/19  \\\n",
       "0  17600.0  30800.0  8800.0        0  17600.0  ...  13288.0  13200.0   2904.0   \n",
       "1        0        0       0        0        0  ...        0        0        0   \n",
       "2        0        0  8800.0  13200.0   8800.0  ...  17600.0  17600.0  17600.0   \n",
       "3        0        0       0        0        0  ...        0        0        0   \n",
       "\n",
       "1 12/5/19 12/16/19 12/16/19  12/16/19  12/4/19 12/4/19 12/4/19  \n",
       "0   352.0        0     88.0         0   2904.0   704.0       0  \n",
       "1       0        0        0         0        0       0       0  \n",
       "2  8800.0  22000.0  22000.0  110000.0  22000.0       0       0  \n",
       "3       0        0        0         0        0       0       0  \n",
       "\n",
       "[4 rows x 244 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "totals = lis.groupby('Phylum', as_index=False, sort=False).sum()\n",
    "totals\n",
    "\n",
    "# empty text-containing columns\n",
    "totals = totals.drop(\"Status\", axis=1)\n",
    "totals.insert(0, 'Status', \"\") \n",
    "totals[\"Genus\"] = \"\"\n",
    "totals[\"Species\"] = \"\"\n",
    "\n",
    "# rename to TOTAL \"   \"\n",
    "totals[\"Phylum\"] = totals[\"Phylum\"].str.upper().apply(lambda x: \"TOTAL \" + x + \"S\")\n",
    "\n",
    "# add in line skips\n",
    "totals = totals.set_index(lis.groupby(['Phylum']).tail(1).index + 0.1)\n",
    "empty_df = pd.DataFrame(\"\", index=lis.groupby(['Phylum']).tail(1).index+0.2, columns=totals.columns)\n",
    "totals = pd.concat([totals, empty_df]).sort_index()\n",
    "totals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ae9e2fd7-6d29-48af-8cd5-64a92cea7d99",
   "metadata": {},
   "outputs": [],
   "source": [
    "lis = pd.concat([lis, totals]).sort_index().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bd68974b-347e-42c9-8789-3acd73f94b38",
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace Nans with zero\n",
    "lis = lis.fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "582eaf09-676e-4c98-b2b9-db8fc2164efb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add back multiheader\n",
    "needed_cols = pd.Series(np.full(len(lis.columns) - len(original_headers), None)) \n",
    "original_headers = pd.concat([needed_cols, original_headers.to_series()], ignore_index=True)\n",
    "lis.columns = pd.MultiIndex.from_arrays([original_headers, lis.columns])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "cf9f1943-4fc4-49da-9f96-508a8842ac0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save dataframe to excel\n",
    "lis.to_excel(f\"outputs/{csv_name}-{str(datetime.now())}.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3041b0fb-f98c-4216-9f3a-ae099e1e6d2f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
