{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c1bb888c-cd9d-4e9b-87de-1432760e5a9e",
   "metadata": {},
   "source": [
    "### HARD CODED RULES\n",
    "* requires format of skipped line between different phylums (ex. empty row above Diatom, Dinoflagellate, etc.)\n",
    "* first word of species name is the genus\n",
    "\n",
    "1. diatoms are NOT mixotrophs\n",
    "2. remove all \"[name]-like\" (without genus specified)\n",
    "3. remove all \"[genus name] spp.\" AND \"[genus name] sp.\" EXCEPT for Ochromonas genus\n",
    "4. assume everything after \"Unknown flagellates\" is irrelevant (to be deleted)\n",
    "\n",
    "Status Key--  \n",
    "Confirmed := explicitly in the Mixotroph Database  \n",
    "Unsure (sp. in mdb) := genus in Mixotroph Database lists \"[genus name] sp.\" (ex. Ochromonas sp. for Ochromonas danica)\n",
    "Unsure (inexact name):= LIS name is in a longer Mixotroph Database name (ex. Chattonella marina in Chattonella marina var. ovata)   "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2718b71-fbb0-4c73-8972-e0dfaa85136c",
   "metadata": {},
   "source": [
    "### QUESTIONS TO ASK\n",
    "\n",
    "1. Should I be considering \"cysts of Linggulodinium polyedrum\" mixotrophs?\n",
    "2. How should I handle these \"unsure\" cases? - see status key above"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae5e8ced-e54e-4c9a-9765-2d4f96e750f5",
   "metadata": {},
   "source": [
    "### Code Steps\n",
    "For Mixotroph Database:\n",
    "1. reset headers\n",
    "2. consider potential speed ups such as phylum to row numbers dictionary\n",
    "\n",
    "For LIS Dataframe:\n",
    "1. save initial header\n",
    "2. reset headers\n",
    "3. delete \"unknown flagellates\" and everything below\n",
    "4. delete any row whose first column contains \"TOTAL\"\n",
    "5. get array of indices of missing values in second column (skipped rows)\n",
    "6. add 1 to every value in that array (to get the indicies of where the phlyum is)\n",
    "7. copy the values at those indicies in the first column to a new first column called phylum\n",
    "8. and delete them where they were before\n",
    "9. backfill the phylum column so that it is completely filled (fill in empty rows using previous value)\n",
    "10. delete all rows where there are missing values in the species name column\n",
    "11. rename what is now the second column as genus\n",
    "12. proceed to clear out what you know it is not based on the hard coded rules:\n",
    "13. delete all rows where phylum is diatom\n",
    "14. delete any row whos value in the species column ends with \"-like\"\n",
    "15. find all rows whose species name contains spp or sp EXCEPT if it also contains ochromonas and delete those rows\n",
    "16. create new column called status to become the first column\n",
    "17. now proceed iteratively through the condensed dataframe:\n",
    "18. for each row:\n",
    "19. check if name is in mixotroph database, if yes, status = confirmed\n",
    "20. if not, is the only name for it in the database sp., if yes, status = unsure (sp. in mdb)\n",
    "21. if not, is there a longer name for it in the database, if yes, status = unsure (inexact name)\n",
    "22. then after iterating:\n",
    "23. for each phylum block, insert a row \"Totals\" and get the sums of everywhere that is that phylum\n",
    "24. save this new file as a csv in the outputs folder and have the name be the old name + the date (timestamp)\n",
    "25. add back initial header combined with current as a multiheader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f0081d70-60c2-4062-a222-7fcfb5fbe6d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "5a6f3687-f38c-4db2-95d2-cdd07727f456",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>1</th>\n",
       "      <th>Species Name</th>\n",
       "      <th>Taxonomic Group</th>\n",
       "      <th>AphiaID</th>\n",
       "      <th>Additional notes</th>\n",
       "      <th>Gene markers</th>\n",
       "      <th>PR2 Accession Number</th>\n",
       "      <th>GenBank Accession Number</th>\n",
       "      <th>Reference to sequence</th>\n",
       "      <th>MFT</th>\n",
       "      <th>Evidence of mixoplankton activity</th>\n",
       "      <th>...</th>\n",
       "      <th>REDS</th>\n",
       "      <th>SANT</th>\n",
       "      <th>SARC</th>\n",
       "      <th>SATL</th>\n",
       "      <th>SPSG</th>\n",
       "      <th>SSTC</th>\n",
       "      <th>SUND</th>\n",
       "      <th>TASM</th>\n",
       "      <th>WARM</th>\n",
       "      <th>WTRA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acanthochiasma sp</td>\n",
       "      <td>Radiolaria</td>\n",
       "      <td>368427</td>\n",
       "      <td>Acantharia</td>\n",
       "      <td>18S_rRNA_nucleus;18S_rRNA_nucleus;18S_rRNA_nuc...</td>\n",
       "      <td>HM103395.1.1099_U;HM103418.1.1104_U;JN811207.1...</td>\n",
       "      <td>HM103395;HM103418;JN811207;GU825020;HM103399;H...</td>\n",
       "      <td>Quaiser,A.. Comparative metagenomics of bathyp...</td>\n",
       "      <td>eSNCM</td>\n",
       "      <td>endosymbionts</td>\n",
       "      <td>...</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acanthometra fusca</td>\n",
       "      <td>Radiolaria</td>\n",
       "      <td>not registered</td>\n",
       "      <td>Acantharia</td>\n",
       "      <td>18S_rRNA_nucleus;18S_rRNA_nucleus;18S_rRNA_nuc...</td>\n",
       "      <td>KC172856.1.1696_U;EU446351.1.1552_U;JN811165.1...</td>\n",
       "      <td>KC172856;EU446351;JN811165</td>\n",
       "      <td>Decelle,J.. Diversity, ecology and biogeochemi...</td>\n",
       "      <td>eSNCM</td>\n",
       "      <td>endosymbionts</td>\n",
       "      <td>...</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Acanthodesmia vinculata</td>\n",
       "      <td>Radiolaria</td>\n",
       "      <td>493675</td>\n",
       "      <td>Acantharia</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>eSNCM</td>\n",
       "      <td>endosymbionts</td>\n",
       "      <td>...</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>15</td>\n",
       "      <td>not recorded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acanthometra pellucida</td>\n",
       "      <td>Radiolaria</td>\n",
       "      <td>235750</td>\n",
       "      <td>Acantharia</td>\n",
       "      <td>18S_rRNA_nucleus;18S_rRNA_nucleus;18S_rRNA_nuc...</td>\n",
       "      <td>JN811196.1.1668_U;JQ697712.1.1693_U;JQ697708.1...</td>\n",
       "      <td>JN811196;JQ697712;JQ697708;JN811190;JQ697711;J...</td>\n",
       "      <td>Decelle,J.. Molecular Phylogeny and Morphologi...</td>\n",
       "      <td>eSNCM</td>\n",
       "      <td>endosymbionts</td>\n",
       "      <td>...</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acanthometron sp.</td>\n",
       "      <td>Radiolaria</td>\n",
       "      <td>391880</td>\n",
       "      <td>Acantharia</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>eSNCM</td>\n",
       "      <td>endosymbionts</td>\n",
       "      <td>...</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 107 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "1             Species Name Taxonomic Group         AphiaID Additional notes  \\\n",
       "0        Acanthochiasma sp      Radiolaria          368427       Acantharia   \n",
       "1       Acanthometra fusca      Radiolaria  not registered       Acantharia   \n",
       "2  Acanthodesmia vinculata      Radiolaria          493675       Acantharia   \n",
       "3   Acanthometra pellucida      Radiolaria          235750       Acantharia   \n",
       "4        Acanthometron sp.      Radiolaria          391880       Acantharia   \n",
       "\n",
       "1                                       Gene markers  \\\n",
       "0  18S_rRNA_nucleus;18S_rRNA_nucleus;18S_rRNA_nuc...   \n",
       "1  18S_rRNA_nucleus;18S_rRNA_nucleus;18S_rRNA_nuc...   \n",
       "2                                       not recorded   \n",
       "3  18S_rRNA_nucleus;18S_rRNA_nucleus;18S_rRNA_nuc...   \n",
       "4                                       not recorded   \n",
       "\n",
       "1                               PR2 Accession Number  \\\n",
       "0  HM103395.1.1099_U;HM103418.1.1104_U;JN811207.1...   \n",
       "1  KC172856.1.1696_U;EU446351.1.1552_U;JN811165.1...   \n",
       "2                                       not recorded   \n",
       "3  JN811196.1.1668_U;JQ697712.1.1693_U;JQ697708.1...   \n",
       "4                                       not recorded   \n",
       "\n",
       "1                           GenBank Accession Number  \\\n",
       "0  HM103395;HM103418;JN811207;GU825020;HM103399;H...   \n",
       "1                         KC172856;EU446351;JN811165   \n",
       "2                                       not recorded   \n",
       "3  JN811196;JQ697712;JQ697708;JN811190;JQ697711;J...   \n",
       "4                                       not recorded   \n",
       "\n",
       "1                              Reference to sequence    MFT  \\\n",
       "0  Quaiser,A.. Comparative metagenomics of bathyp...  eSNCM   \n",
       "1  Decelle,J.. Diversity, ecology and biogeochemi...  eSNCM   \n",
       "2                                       not recorded  eSNCM   \n",
       "3  Decelle,J.. Molecular Phylogeny and Morphologi...  eSNCM   \n",
       "4                                       not recorded  eSNCM   \n",
       "\n",
       "1 Evidence of mixoplankton activity  ...          REDS          SANT  \\\n",
       "0                     endosymbionts  ...  not recorded  not recorded   \n",
       "1                     endosymbionts  ...  not recorded  not recorded   \n",
       "2                     endosymbionts  ...  not recorded  not recorded   \n",
       "3                     endosymbionts  ...  not recorded  not recorded   \n",
       "4                     endosymbionts  ...  not recorded  not recorded   \n",
       "\n",
       "1          SARC          SATL          SPSG          SSTC          SUND  \\\n",
       "0  not recorded  not recorded  not recorded  not recorded  not recorded   \n",
       "1  not recorded  not recorded  not recorded  not recorded  not recorded   \n",
       "2  not recorded             2             8  not recorded  not recorded   \n",
       "3  not recorded  not recorded  not recorded  not recorded  not recorded   \n",
       "4  not recorded  not recorded  not recorded  not recorded  not recorded   \n",
       "\n",
       "1          TASM          WARM          WTRA  \n",
       "0  not recorded  not recorded  not recorded  \n",
       "1  not recorded  not recorded  not recorded  \n",
       "2  not recorded            15  not recorded  \n",
       "3  not recorded  not recorded  not recorded  \n",
       "4  not recorded  not recorded  not recorded  \n",
       "\n",
       "[5 rows x 107 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdb = pd.read_csv(\"csvs/MDB - 3Dec2022.csv\")\n",
    "mdb.columns = mdb.iloc[1]\n",
    "mdb = mdb.drop([0, 1]).reset_index(drop=True)\n",
    "mdb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e852c70f-e894-4588-bc06-4a3079f32083",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>1</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Species</th>\n",
       "      <th>1/3/19</th>\n",
       "      <th>1/3/19</th>\n",
       "      <th>1/3/19</th>\n",
       "      <th>1/7/19</th>\n",
       "      <th>1/7/19</th>\n",
       "      <th>1/7/19</th>\n",
       "      <th>1/2/19</th>\n",
       "      <th>1/2/19</th>\n",
       "      <th>...</th>\n",
       "      <th>12/6/19</th>\n",
       "      <th>12/6/19</th>\n",
       "      <th>12/6/19</th>\n",
       "      <th>12/5/19</th>\n",
       "      <th>12/16/19</th>\n",
       "      <th>12/16/19</th>\n",
       "      <th>12/16/19</th>\n",
       "      <th>12/4/19</th>\n",
       "      <th>12/4/19</th>\n",
       "      <th>12/4/19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Diatom</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Achnanthes</td>\n",
       "      <td>Achnanthes spp.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>176</td>\n",
       "      <td>NaN</td>\n",
       "      <td>88</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Actinocyclus</td>\n",
       "      <td>Actinocyclus spp.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Actinoptychus</td>\n",
       "      <td>Actinoptychus senarius</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 242 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "1         Phylum                 Species 1/3/19 1/3/19 1/3/19 1/7/19 1/7/19  \\\n",
       "0            NaN                     NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "1         Diatom                     NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "2     Achnanthes         Achnanthes spp.    NaN    352    NaN    NaN    NaN   \n",
       "3   Actinocyclus       Actinocyclus spp.    NaN    NaN    NaN    NaN    NaN   \n",
       "4  Actinoptychus  Actinoptychus senarius    NaN    NaN    NaN    NaN    NaN   \n",
       "\n",
       "1 1/7/19 1/2/19 1/2/19  ... 12/6/19 12/6/19 12/6/19 12/5/19 12/16/19 12/16/19  \\\n",
       "0    NaN    NaN    NaN  ...     NaN     NaN     NaN     NaN      NaN      NaN   \n",
       "1    NaN    NaN    NaN  ...     NaN     NaN     NaN     NaN      NaN      NaN   \n",
       "2    NaN    NaN    NaN  ...     176     NaN      88     NaN      NaN      NaN   \n",
       "3    NaN    NaN    NaN  ...     NaN     NaN     NaN     NaN      NaN      NaN   \n",
       "4    NaN    NaN    NaN  ...     NaN     NaN     NaN     NaN      NaN      NaN   \n",
       "\n",
       "1 12/16/19 12/4/19 12/4/19 12/4/19  \n",
       "0      NaN     NaN     NaN     NaN  \n",
       "1      NaN     NaN     NaN     NaN  \n",
       "2      NaN     NaN     NaN     NaN  \n",
       "3      NaN     NaN     NaN     NaN  \n",
       "4      NaN     NaN     NaN     NaN  \n",
       "\n",
       "[5 rows x 242 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lis = pd.read_csv(\"csvs/LIS_2019-Phytoplankton_Final Report Data.csv\")\n",
    "lis.columns = lis.iloc[1]\n",
    "lis = lis.iloc[3:].reset_index(drop=True)\n",
    "lis.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98fc39f9-f6db-4d47-afe8-753c0bf1294a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
