{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c1bb888c-cd9d-4e9b-87de-1432760e5a9e",
   "metadata": {},
   "source": [
    "### HARD CODED RULES\n",
    "Requires format of skipped line between different phylums (ex. empty row above Diatom, Dinoflagellate, etc.)\n",
    "\n",
    "Assumed all {Ochromonas, } are mixotrophs.\n",
    "\n",
    "1. assume everything after \"Unknown flagellates\" is irrelevant (to be deleted)\n",
    "2. diatoms are NOT mixotrophs\n",
    "3. remove all \"[name]-like\" (without genus specified)\n",
    "4. remove all \"[genus name] spp.\" AND \"[genus name] sp.\"\n",
    "5. check \"cysts of\"\n",
    "\n",
    "Status Key--  \n",
    "Confirmed := explicitly in the Mixotroph Database  \n",
    "Unsure (sp. in mdb) := genus in Mixotroph Database lists \"[genus name] sp.\" (ex. Ochromonas sp. for Ochromonas danica)  \n",
    "Unsure (inexact name):= LIS name is in a longer Mixotroph Database name or vice versa (ex. Chattonella marina in Chattonella marina var. ovata)   "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2718b71-fbb0-4c73-8972-e0dfaa85136c",
   "metadata": {},
   "source": [
    "### QUESTIONS TO ASK\n",
    "\n",
    "1. Should I be considering \"cysts of Linggulodinium polyedrum\" mixotrophs?\n",
    "2. How should I handle these \"unsure\" cases? - see status key above"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae5e8ced-e54e-4c9a-9765-2d4f96e750f5",
   "metadata": {},
   "source": [
    "### Code Steps\n",
    "For Mixotroph Database:\n",
    "1. reset headers\n",
    "2. consider potential speed ups such as phylum to row numbers dictionary\n",
    "\n",
    "For LIS Dataframe:\n",
    "1. save initial header\n",
    "2. reset headers\n",
    "3. delete \"unknown flagellates\" and everything below\n",
    "4. delete any row whose first column contains \"TOTAL\"\n",
    "5. get array of indices of missing values in second column (skipped rows)\n",
    "6. add 1 to every value in that array (to get the indicies of where the phlyum is)\n",
    "7. copy the values at those indicies in the first column to a new first column called phylum\n",
    "8. and delete them where they were before\n",
    "9. backfill the phylum column so that it is completely filled (fill in empty rows using previous value)\n",
    "10. delete all rows where there are missing values in the species name column\n",
    "11. rename what is now the second column as genus\n",
    "12. proceed to clear out what you know it is not based on the hard coded rules:\n",
    "13. delete all rows where phylum is diatom\n",
    "14. delete any row whos value in the species column ends with \"-like\"\n",
    "15. find all rows whose species name contains spp or sp EXCEPT if it also contains ochromonas and delete those rows\n",
    "16. create new column called status to become the first column\n",
    "17. now proceed iteratively through the condensed dataframe:\n",
    "18. check if name is in mixotroph database, if yes, status = confirmed\n",
    "19. if not, is the only name for it in the database sp., if yes, status = unsure (sp. in mdb)\n",
    "20. if not, is there a longer name for it in the database, if yes, status = unsure (inexact name)\n",
    "21. drop all rows with status \"None\"\n",
    "22. for each phylum block, insert a row \"Totals\" and get the sums of everywhere that is that phylum\n",
    "23. save this new file as a csv in the outputs folder and have the name be the old name + the date (timestamp)\n",
    "24. add back initial header combined with current as a multiheader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "id": "f0081d70-60c2-4062-a222-7fcfb5fbe6d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "id": "5a6f3687-f38c-4db2-95d2-cdd07727f456",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>1</th>\n",
       "      <th>Species Name</th>\n",
       "      <th>Taxonomic Group</th>\n",
       "      <th>AphiaID</th>\n",
       "      <th>Additional notes</th>\n",
       "      <th>Gene markers</th>\n",
       "      <th>PR2 Accession Number</th>\n",
       "      <th>GenBank Accession Number</th>\n",
       "      <th>Reference to sequence</th>\n",
       "      <th>MFT</th>\n",
       "      <th>Evidence of mixoplankton activity</th>\n",
       "      <th>...</th>\n",
       "      <th>REDS</th>\n",
       "      <th>SANT</th>\n",
       "      <th>SARC</th>\n",
       "      <th>SATL</th>\n",
       "      <th>SPSG</th>\n",
       "      <th>SSTC</th>\n",
       "      <th>SUND</th>\n",
       "      <th>TASM</th>\n",
       "      <th>WARM</th>\n",
       "      <th>WTRA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acanthochiasma sp</td>\n",
       "      <td>Radiolaria</td>\n",
       "      <td>368427</td>\n",
       "      <td>Acantharia</td>\n",
       "      <td>18S_rRNA_nucleus;18S_rRNA_nucleus;18S_rRNA_nuc...</td>\n",
       "      <td>HM103395.1.1099_U;HM103418.1.1104_U;JN811207.1...</td>\n",
       "      <td>HM103395;HM103418;JN811207;GU825020;HM103399;H...</td>\n",
       "      <td>Quaiser,A.. Comparative metagenomics of bathyp...</td>\n",
       "      <td>eSNCM</td>\n",
       "      <td>endosymbionts</td>\n",
       "      <td>...</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acanthometra fusca</td>\n",
       "      <td>Radiolaria</td>\n",
       "      <td>not registered</td>\n",
       "      <td>Acantharia</td>\n",
       "      <td>18S_rRNA_nucleus;18S_rRNA_nucleus;18S_rRNA_nuc...</td>\n",
       "      <td>KC172856.1.1696_U;EU446351.1.1552_U;JN811165.1...</td>\n",
       "      <td>KC172856;EU446351;JN811165</td>\n",
       "      <td>Decelle,J.. Diversity, ecology and biogeochemi...</td>\n",
       "      <td>eSNCM</td>\n",
       "      <td>endosymbionts</td>\n",
       "      <td>...</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Acanthodesmia vinculata</td>\n",
       "      <td>Radiolaria</td>\n",
       "      <td>493675</td>\n",
       "      <td>Acantharia</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>eSNCM</td>\n",
       "      <td>endosymbionts</td>\n",
       "      <td>...</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>15</td>\n",
       "      <td>not recorded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acanthometra pellucida</td>\n",
       "      <td>Radiolaria</td>\n",
       "      <td>235750</td>\n",
       "      <td>Acantharia</td>\n",
       "      <td>18S_rRNA_nucleus;18S_rRNA_nucleus;18S_rRNA_nuc...</td>\n",
       "      <td>JN811196.1.1668_U;JQ697712.1.1693_U;JQ697708.1...</td>\n",
       "      <td>JN811196;JQ697712;JQ697708;JN811190;JQ697711;J...</td>\n",
       "      <td>Decelle,J.. Molecular Phylogeny and Morphologi...</td>\n",
       "      <td>eSNCM</td>\n",
       "      <td>endosymbionts</td>\n",
       "      <td>...</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Acanthometron sp.</td>\n",
       "      <td>Radiolaria</td>\n",
       "      <td>391880</td>\n",
       "      <td>Acantharia</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>eSNCM</td>\n",
       "      <td>endosymbionts</td>\n",
       "      <td>...</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "      <td>not recorded</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 107 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "1             Species Name Taxonomic Group         AphiaID Additional notes  \\\n",
       "0        Acanthochiasma sp      Radiolaria          368427       Acantharia   \n",
       "1       Acanthometra fusca      Radiolaria  not registered       Acantharia   \n",
       "2  Acanthodesmia vinculata      Radiolaria          493675       Acantharia   \n",
       "3   Acanthometra pellucida      Radiolaria          235750       Acantharia   \n",
       "4        Acanthometron sp.      Radiolaria          391880       Acantharia   \n",
       "\n",
       "1                                       Gene markers  \\\n",
       "0  18S_rRNA_nucleus;18S_rRNA_nucleus;18S_rRNA_nuc...   \n",
       "1  18S_rRNA_nucleus;18S_rRNA_nucleus;18S_rRNA_nuc...   \n",
       "2                                       not recorded   \n",
       "3  18S_rRNA_nucleus;18S_rRNA_nucleus;18S_rRNA_nuc...   \n",
       "4                                       not recorded   \n",
       "\n",
       "1                               PR2 Accession Number  \\\n",
       "0  HM103395.1.1099_U;HM103418.1.1104_U;JN811207.1...   \n",
       "1  KC172856.1.1696_U;EU446351.1.1552_U;JN811165.1...   \n",
       "2                                       not recorded   \n",
       "3  JN811196.1.1668_U;JQ697712.1.1693_U;JQ697708.1...   \n",
       "4                                       not recorded   \n",
       "\n",
       "1                           GenBank Accession Number  \\\n",
       "0  HM103395;HM103418;JN811207;GU825020;HM103399;H...   \n",
       "1                         KC172856;EU446351;JN811165   \n",
       "2                                       not recorded   \n",
       "3  JN811196;JQ697712;JQ697708;JN811190;JQ697711;J...   \n",
       "4                                       not recorded   \n",
       "\n",
       "1                              Reference to sequence    MFT  \\\n",
       "0  Quaiser,A.. Comparative metagenomics of bathyp...  eSNCM   \n",
       "1  Decelle,J.. Diversity, ecology and biogeochemi...  eSNCM   \n",
       "2                                       not recorded  eSNCM   \n",
       "3  Decelle,J.. Molecular Phylogeny and Morphologi...  eSNCM   \n",
       "4                                       not recorded  eSNCM   \n",
       "\n",
       "1 Evidence of mixoplankton activity  ...          REDS          SANT  \\\n",
       "0                     endosymbionts  ...  not recorded  not recorded   \n",
       "1                     endosymbionts  ...  not recorded  not recorded   \n",
       "2                     endosymbionts  ...  not recorded  not recorded   \n",
       "3                     endosymbionts  ...  not recorded  not recorded   \n",
       "4                     endosymbionts  ...  not recorded  not recorded   \n",
       "\n",
       "1          SARC          SATL          SPSG          SSTC          SUND  \\\n",
       "0  not recorded  not recorded  not recorded  not recorded  not recorded   \n",
       "1  not recorded  not recorded  not recorded  not recorded  not recorded   \n",
       "2  not recorded             2             8  not recorded  not recorded   \n",
       "3  not recorded  not recorded  not recorded  not recorded  not recorded   \n",
       "4  not recorded  not recorded  not recorded  not recorded  not recorded   \n",
       "\n",
       "1          TASM          WARM          WTRA  \n",
       "0  not recorded  not recorded  not recorded  \n",
       "1  not recorded  not recorded  not recorded  \n",
       "2  not recorded            15  not recorded  \n",
       "3  not recorded  not recorded  not recorded  \n",
       "4  not recorded  not recorded  not recorded  \n",
       "\n",
       "[5 rows x 107 columns]"
      ]
     },
     "execution_count": 272,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mdb = pd.read_csv(\"csvs/MDB - 3Dec2022.csv\")\n",
    "mdb.columns = mdb.iloc[1]\n",
    "mdb = mdb.drop([0, 1]).reset_index(drop=True)\n",
    "mdb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "id": "dd7eb54a-5368-4369-81e5-e5f4d593198f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# edit mdb so that species ending in \"sp\" now end in \"sp.\"\n",
    "mdb['Species Name'] = mdb['Species Name'].str.replace(r'sp$', 'sp.', regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "id": "e852c70f-e894-4588-bc06-4a3079f32083",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import and clean LIS data\n",
    "lis = pd.read_csv(\"csvs/LIS_2019-Phytoplankton_Final Report Data.csv\")\n",
    "original_headers = lis.columns  # save original column headers\n",
    "lis.columns = lis.iloc[1]  # reset column headers\n",
    "lis = lis.iloc[3:].reset_index(drop=True)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 275,
   "id": "ae223ae9-e76a-4383-ad84-b79487293ab6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove rows after unknown flagellates\n",
    "unknown_flagellates_ind = lis[lis[\"Phylum\"] == \"Unknown flagellates\"].index[0] \n",
    "lis = lis.iloc[:unknown_flagellates_ind]\n",
    "lis = lis.iloc[:lis.last_valid_index()+1]  # remove trailing nan rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "id": "948f15e0-5c49-4fd1-af99-bb7af42bdce8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove rows that contain \"TOTAL\"\n",
    "lis = lis[~lis[\"Phylum\"].str.contains(\"TOTAL\", na=False)].reset_index(drop=True)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "id": "28a87a73-ccc3-4a22-bc8e-f22c68ebcbb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# construct correct phylum column\n",
    "actual_phylum_ind = lis[lis[\"Species\"].isna() & lis[\"Phylum\"].isna()].index + 1\n",
    "lis = lis.rename(columns={\"Phylum\": \"Genus\"}) # rename phylum column to genus\n",
    "lis.insert(0, 'Phylum', lis[\"Genus\"].iloc[actual_phylum_ind])  # reconstruct phylum column\n",
    "lis['Phylum'] = lis['Phylum'].ffill()  # forwardfill phylum\n",
    "\n",
    "lis['Genus'] = lis['Species'].str.split().str[0]  # fill genus using first word of species name\n",
    "\n",
    "lis = lis.dropna(subset=['Species']).reset_index(drop=True) # delete rows with na in Species column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "id": "ca4ca7ba-58a3-4493-9055-961238f4182e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add Status column\n",
    "lis.insert(0, 'Status', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "id": "20575f78-e4dc-498b-8ddb-d07d9140aeb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# store blocks of known mixotroph genuses \n",
    "ochromonas_ind = lis[lis[\"Species\"].str.contains(\"Ochromonas\")].index\n",
    "ochromonas_block = lis.iloc[ochromonas_ind] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "id": "f4a71b12-d2b8-4700-a53c-2c17e8aba73b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove based on hard coded rules (NOT RESETTING INDEX IN ORDER TO ADD BLOCKS BACK CORRECTLY)\n",
    "lis = lis[lis[\"Phylum\"] != \"Diatom\"] # remove all diatoms\n",
    "lis = lis[~lis[\"Species\"].str.contains(\"-like\")] # remove species ending with \"-like\"\n",
    "lis = lis[~lis[\"Species\"].str.contains(\"sp.|spp.\")]  # remove all sp. / spp."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "id": "24ce9749-f358-4b2b-956f-f9e94032cf49",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check \"cysts of\"\n",
    "CYSTS_LEN = len(\"cysts of \")\n",
    "cysts_of = lis[lis[\"Species\"].str.contains(\"cysts of\", regex=False)][\"Species\"].str.slice(CYSTS_LEN)\n",
    "filtered = cysts_of.isin(mdb['Species Name'])\n",
    "lis.loc[filtered[filtered].index, \"Status\"] = \"Confirmed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "id": "c635deda-718b-46a5-bd0f-654d7e504f6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add back stored blocks of known mixotrophs and mark as Confirmed\n",
    "lis = pd.concat([lis, ochromonas_block]).sort_index().drop_duplicates()\n",
    "lis.loc[ochromonas_ind, \"Status\"] = \"Confirmed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "id": "a6f4a0bc-3b23-49ce-b3c9-200cd295a498",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check if (in none status) direct match and mark all Trues as \"Confirmed\"\n",
    "filtered = lis[lis['Status'].isnull()][\"Species\"].isin(mdb['Species Name'])\n",
    "lis.loc[filtered[filtered].index, \"Status\"] = \"Confirmed\"\n",
    "\n",
    "# check (in remaining none status) if the genus has sp. and mark all Trues as \"Unsure (sp. in mdb)\"\n",
    "genus_to_check = lis[lis['Status'].isnull()]['Species'].str.split().str[0].drop_duplicates() + \" sp.\"\n",
    "filtered = genus_to_check.isin(mdb['Species Name'])\n",
    "lis.loc[filtered[filtered].index, \"Status\"] = \"Unsure (sp. mdb)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "id": "4d751f4e-d0e8-4ac5-bd3d-a5344d8ce133",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check (in remaining none status) if the name is contained in the mdb and vice versa and mark all Trues as \"Unsure (inexact name)\"\n",
    "filtered = lis[lis['Status'].isnull()][\"Species\"].apply(lambda x: mdb[\"Species Name\"].str.contains(x, regex=False).any())\n",
    "lis.loc[filtered[filtered].index, \"Status\"] = \"Unsure (inexact name)\"\n",
    "\n",
    "pattern = '|'.join(mdb['Species Name'])\n",
    "filtered = lis[lis['Status'].isnull()][\"Species\"].str.contains(pattern, regex=True)\n",
    "lis.loc[filtered[filtered].index, \"Status\"] = \"Unsure (inexact name)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "id": "96679bc3-4754-4fdb-bf9f-90b4d7b1d240",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>1</th>\n",
       "      <th>Status</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>1/3/19</th>\n",
       "      <th>1/3/19</th>\n",
       "      <th>1/3/19</th>\n",
       "      <th>1/7/19</th>\n",
       "      <th>1/7/19</th>\n",
       "      <th>1/7/19</th>\n",
       "      <th>...</th>\n",
       "      <th>12/6/19</th>\n",
       "      <th>12/6/19</th>\n",
       "      <th>12/6/19</th>\n",
       "      <th>12/5/19</th>\n",
       "      <th>12/16/19</th>\n",
       "      <th>12/16/19</th>\n",
       "      <th>12/16/19</th>\n",
       "      <th>12/4/19</th>\n",
       "      <th>12/4/19</th>\n",
       "      <th>12/4/19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Dinoflagellate</td>\n",
       "      <td>Akashiwo</td>\n",
       "      <td>Akashiwo sanguinea</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Dinoflagellate</td>\n",
       "      <td>Dinophysis</td>\n",
       "      <td>Dinophysis acuminata</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Dinoflagellate</td>\n",
       "      <td>Dinophysis</td>\n",
       "      <td>Dinophysis miles</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Dinoflagellate</td>\n",
       "      <td>Dinophysis</td>\n",
       "      <td>Dinophysis norvegica</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Dinoflagellate</td>\n",
       "      <td>Gambierdiscus</td>\n",
       "      <td>Gambierdiscus toxicus</td>\n",
       "      <td>352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>88</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>88</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Dinoflagellate</td>\n",
       "      <td>Gonyaulax</td>\n",
       "      <td>Gonyaulax polygramma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164</th>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Dinoflagellate</td>\n",
       "      <td>Heterocapsa</td>\n",
       "      <td>Heterocapsa circularisquama</td>\n",
       "      <td>17600</td>\n",
       "      <td>17600</td>\n",
       "      <td>30800</td>\n",
       "      <td>8800</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17600</td>\n",
       "      <td>...</td>\n",
       "      <td>13200</td>\n",
       "      <td>13200</td>\n",
       "      <td>2904</td>\n",
       "      <td>352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2904</td>\n",
       "      <td>704</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Dinoflagellate</td>\n",
       "      <td>Noctiluca</td>\n",
       "      <td>Noctiluca scintillans</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Dinoflagellate</td>\n",
       "      <td>Prorocentrum</td>\n",
       "      <td>Prorocentrum lima</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Dinoflagellate</td>\n",
       "      <td>Prorocentrum</td>\n",
       "      <td>Prorocentrum micans</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>Unsure (inexact name)</td>\n",
       "      <td>Raphidophyte</td>\n",
       "      <td>Chattonella</td>\n",
       "      <td>Chattonella marina</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Raphidophyte</td>\n",
       "      <td>Heterosigma</td>\n",
       "      <td>Heterosigma akashiwo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Ochrophyte (diatoms excluded)</td>\n",
       "      <td>Ochromonas</td>\n",
       "      <td>Ochromonas spp.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8800</td>\n",
       "      <td>13200</td>\n",
       "      <td>8800</td>\n",
       "      <td>...</td>\n",
       "      <td>17600</td>\n",
       "      <td>17600</td>\n",
       "      <td>17600</td>\n",
       "      <td>8800</td>\n",
       "      <td>22000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>110000</td>\n",
       "      <td>22000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Ochrophyte (diatoms excluded)</td>\n",
       "      <td>Ochromonas</td>\n",
       "      <td>Ochromonas danica</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>Confirmed</td>\n",
       "      <td>Haptophyte</td>\n",
       "      <td>Phaeocystis</td>\n",
       "      <td>Phaeocystis globosa</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15 rows × 244 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "1                   Status                         Phylum          Genus  \\\n",
       "143              Confirmed                 Dinoflagellate       Akashiwo   \n",
       "152              Confirmed                 Dinoflagellate     Dinophysis   \n",
       "153              Confirmed                 Dinoflagellate     Dinophysis   \n",
       "154              Confirmed                 Dinoflagellate     Dinophysis   \n",
       "156              Confirmed                 Dinoflagellate  Gambierdiscus   \n",
       "158              Confirmed                 Dinoflagellate      Gonyaulax   \n",
       "164              Confirmed                 Dinoflagellate    Heterocapsa   \n",
       "167              Confirmed                 Dinoflagellate      Noctiluca   \n",
       "180              Confirmed                 Dinoflagellate   Prorocentrum   \n",
       "182              Confirmed                 Dinoflagellate   Prorocentrum   \n",
       "200  Unsure (inexact name)                   Raphidophyte    Chattonella   \n",
       "203              Confirmed                   Raphidophyte    Heterosigma   \n",
       "212              Confirmed  Ochrophyte (diatoms excluded)     Ochromonas   \n",
       "213              Confirmed  Ochrophyte (diatoms excluded)     Ochromonas   \n",
       "216              Confirmed                     Haptophyte    Phaeocystis   \n",
       "\n",
       "1                        Species 1/3/19 1/3/19 1/3/19 1/7/19 1/7/19 1/7/19  \\\n",
       "143           Akashiwo sanguinea    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "152         Dinophysis acuminata    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "153             Dinophysis miles    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "154         Dinophysis norvegica    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "156        Gambierdiscus toxicus    352    NaN    NaN    NaN    NaN    NaN   \n",
       "158         Gonyaulax polygramma    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "164  Heterocapsa circularisquama  17600  17600  30800   8800    NaN  17600   \n",
       "167        Noctiluca scintillans    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "180            Prorocentrum lima    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "182          Prorocentrum micans    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "200           Chattonella marina    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "203         Heterosigma akashiwo    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "212              Ochromonas spp.    NaN    NaN    NaN   8800  13200   8800   \n",
       "213            Ochromonas danica    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "216          Phaeocystis globosa    NaN    NaN    NaN    NaN    NaN    NaN   \n",
       "\n",
       "1    ... 12/6/19 12/6/19 12/6/19 12/5/19 12/16/19 12/16/19 12/16/19 12/4/19  \\\n",
       "143  ...     NaN     NaN     NaN     NaN      NaN      NaN      NaN     NaN   \n",
       "152  ...     NaN     NaN     NaN     NaN      NaN      NaN      NaN     NaN   \n",
       "153  ...     NaN     NaN     NaN     NaN      NaN      NaN      NaN     NaN   \n",
       "154  ...     NaN     NaN     NaN     NaN      NaN      NaN      NaN     NaN   \n",
       "156  ...      88     NaN     NaN     NaN      NaN       88      NaN     NaN   \n",
       "158  ...     NaN     NaN     NaN     NaN      NaN      NaN      NaN     NaN   \n",
       "164  ...   13200   13200    2904     352      NaN      NaN      NaN    2904   \n",
       "167  ...     NaN     NaN     NaN     NaN      NaN      NaN      NaN     NaN   \n",
       "180  ...     NaN     NaN     NaN     NaN      NaN      NaN      NaN     NaN   \n",
       "182  ...     NaN     NaN     NaN     NaN      NaN      NaN      NaN     NaN   \n",
       "200  ...     NaN     NaN     NaN     NaN      NaN      NaN      NaN     NaN   \n",
       "203  ...     NaN     NaN     NaN     NaN      NaN      NaN      NaN     NaN   \n",
       "212  ...   17600   17600   17600    8800    22000      NaN   110000   22000   \n",
       "213  ...     NaN     NaN     NaN     NaN      NaN    22000      NaN     NaN   \n",
       "216  ...     NaN     NaN     NaN     NaN      NaN      NaN      NaN     NaN   \n",
       "\n",
       "1   12/4/19 12/4/19  \n",
       "143     NaN     NaN  \n",
       "152     NaN     NaN  \n",
       "153     NaN     NaN  \n",
       "154     NaN     NaN  \n",
       "156     NaN     NaN  \n",
       "158     NaN     NaN  \n",
       "164     704     NaN  \n",
       "167     NaN     NaN  \n",
       "180     NaN     NaN  \n",
       "182     NaN     NaN  \n",
       "200     NaN     NaN  \n",
       "203     NaN     NaN  \n",
       "212     NaN     NaN  \n",
       "213     NaN     NaN  \n",
       "216     NaN     NaN  \n",
       "\n",
       "[15 rows x 244 columns]"
      ]
     },
     "execution_count": 285,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop all rows with Status = \"None\"\n",
    "lis = lis.dropna(subset=['Status'])\n",
    "lis"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
